File: codes/python/ch13/bc_vibrating_string.py
#!/usr/bin/env python3
"""
bc_vibrating_string.py
Chapter 13: Boundary Conditions in Spectral Methods

Vibrating string with a free end -- eigenvalue problem:

    -u_xx = lambda u,   x in [0, 1]
    u(0) = 0            (Dirichlet, fixed end)
    u'(1) = 0           (Neumann, free end)

Exact eigenvalues:    lambda_n = ((2n - 1) pi / 2)^2,  n = 1, 2, ...
Exact eigenfunctions: u_n(x) = sin((2n - 1) pi x / 2)

The Chebyshev grid on [-1, 1] is mapped to [0, 1] via x_phys = (1 + x_cheb)/2,
so that x_cheb = -1 maps to x_phys = 0 and x_cheb = 1 maps to x_phys = 1.
The derivative matrix is scaled accordingly: D_phys = 2 D_cheb.

Boundary conditions are imposed via boundary bordering:
    Row 0  (x_cheb = 1, x_phys = 1):  D_phys[0,:] (Neumann u'(1) = 0)
    Row N  (x_cheb = -1, x_phys = 0): identity row (Dirichlet u(0) = 0)

Generates Figure 13.7: eigenmodes and eigenvalue errors.

Author: Dr. Denys Dutykh
        Mathematics Department
        Khalifa University of Science and Technology
        Abu Dhabi, UAE

Part of the book "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""

import numpy as np
import scipy.linalg
import matplotlib.pyplot as plt
from matplotlib import rcParams
from pathlib import Path

# Import Chebyshev matrix function from Chapter 7
import sys
sys.path.insert(0, str(Path(__file__).parent.parent / 'ch07'))
from cheb_matrix import cheb_matrix

# -----------------------------------------------------------------------------
# Publication-quality matplotlib configuration
# -----------------------------------------------------------------------------
rcParams['font.family'] = 'serif'
rcParams['font.serif'] = ['Computer Modern Roman', 'CMU Serif', 'DejaVu Serif']
rcParams['mathtext.fontset'] = 'cm'
rcParams['axes.labelsize'] = 11
rcParams['axes.titlesize'] = 12
rcParams['xtick.labelsize'] = 10
rcParams['ytick.labelsize'] = 10
rcParams['legend.fontsize'] = 10
rcParams['figure.dpi'] = 150
rcParams['savefig.dpi'] = 300
rcParams['text.usetex'] = False
rcParams['axes.linewidth'] = 0.8
rcParams['xtick.major.width'] = 0.8
rcParams['ytick.major.width'] = 0.8

# -----------------------------------------------------------------------------
# Book color scheme
# -----------------------------------------------------------------------------
NAVY = '#142D6E'
SKY = '#7896D2'
CORAL = '#E74C3C'
TEAL = '#16A085'
PURPLE = '#8E44AD'
ORANGE = '#E67E22'

# Output paths
SCRIPT_DIR = Path(__file__).parent
OUTPUT_DIR = SCRIPT_DIR / '..' / '..' / '..' / 'textbook' / 'figures' / 'ch13' / 'python'


# -----------------------------------------------------------------------------
# Exact solutions
# -----------------------------------------------------------------------------
def exact_eigenvalue(n):
    """Exact eigenvalue lambda_n = ((2n-1) pi / 2)^2 for n = 1, 2, ..."""
    return ((2 * n - 1) * np.pi / 2) ** 2


def exact_eigenfunction(x, n):
    """Exact eigenfunction u_n(x) = sin((2n-1) pi x / 2)."""
    return np.sin((2 * n - 1) * np.pi * x / 2)


# -----------------------------------------------------------------------------
# Solver
# -----------------------------------------------------------------------------
def solve_vibrating_string(N):
    """
    Solve -u_xx = lambda u on [0,1] with u(0)=0, u'(1)=0.

    Uses Chebyshev collocation with boundary bordering.  The eigenvalue
    problem is formulated as a generalised eigenproblem A u = lambda B u
    so that the boundary rows (which do not involve lambda) can be
    encoded by setting the corresponding rows of B to zero.

    Layout (Chebyshev ordering: x_0 = 1 -> x_N = -1):
        - Row 0  (x_phys = 1): Neumann  u'(1) = 0    -> A[0,:] = D, B[0,:] = 0
        - Row N  (x_phys = 0): Dirichlet u(0) = 0     -> A[N,:] = e_N, B[N,:] = 0
        - Interior rows:       -u'' = lambda u         -> A = -D2, B = I

    Parameters
    ----------
    N : int
        Number of Chebyshev intervals.

    Returns
    -------
    eigenvalues : ndarray
        Computed eigenvalues (sorted, positive, real).
    eigenvectors : ndarray, shape (N+1, ?)
        Corresponding eigenvectors on the physical grid.
    x_phys : ndarray, shape (N+1,)
        Physical grid points on [0, 1].
    """
    n = N + 1

    # Chebyshev grid and derivative on [-1, 1]
    D_cheb, x_cheb = cheb_matrix(N)

    # Map to [0, 1]:  x_phys = (1 + x_cheb) / 2
    x_phys = (1.0 + x_cheb) / 2.0
    D_phys = 2.0 * D_cheb          # d/dx_phys = 2 * d/dx_cheb
    D2_phys = D_phys @ D_phys      # second derivative

    # Build generalised eigenproblem  A u = lambda B u
    A = -D2_phys.copy()
    B = np.eye(n)

    # Neumann BC at x_phys = 1 (row 0): u'(1) = 0
    A[0, :] = D_phys[0, :]
    B[0, :] = 0.0

    # Dirichlet BC at x_phys = 0 (row N): u(0) = 0
    A[N, :] = 0.0
    A[N, N] = 1.0
    B[N, :] = 0.0

    # Solve the generalised eigenvalue problem
    eigenvalues, eigenvectors = scipy.linalg.eig(A, B)

    # Extract real parts (eigenvalues should be real for this self-adjoint problem)
    eigenvalues = eigenvalues.real
    eigenvectors = eigenvectors.real

    # Select finite, positive eigenvalues (boundary rows yield inf/nan)
    good_mask = np.isfinite(eigenvalues) & (eigenvalues > 1e-6)
    eigenvalues = eigenvalues[good_mask]
    eigenvectors = eigenvectors[:, good_mask]

    # Sort by eigenvalue
    idx = np.argsort(eigenvalues)
    eigenvalues = eigenvalues[idx]
    eigenvectors = eigenvectors[:, idx]

    return eigenvalues, eigenvectors, x_phys


def main():
    """Create vibrating string eigenvalue figure."""

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    N = 36

    # Solve
    eigenvalues, eigenvectors, x_phys = solve_vibrating_string(N)

    # Fine grid for exact eigenfunctions
    x_fine = np.linspace(0, 1, 500)

    # Number of modes to compare
    n_modes_plot = 6
    n_modes_err = min(len(eigenvalues), N // 2)

    # Colors for the 6 modes
    mode_colors = [NAVY, CORAL, TEAL, PURPLE, ORANGE, SKY]

    # =========================================================================
    # Figure: 2 panels
    # =========================================================================
    fig, axes = plt.subplots(1, 2, figsize=(11, 4.5))

    # ---- Left panel: first 6 eigenmodes ----
    ax1 = axes[0]

    for k in range(n_modes_plot):
        n_mode = k + 1  # mode number (1-indexed)
        lam_exact = exact_eigenvalue(n_mode)

        # Numerical eigenfunction
        u_num = eigenvectors[:, k]

        # Exact eigenfunction on the spectral grid (for sign alignment)
        u_exact_grid = exact_eigenfunction(x_phys, n_mode)

        # Align sign: if inner product is negative, flip
        if np.dot(u_num, u_exact_grid) < 0:
            u_num = -u_num

        # Normalise to unit amplitude
        u_num = u_num / np.max(np.abs(u_num))

        # Exact eigenfunction on fine grid
        u_exact_fine = exact_eigenfunction(x_fine, n_mode)
        u_exact_fine = u_exact_fine / np.max(np.abs(u_exact_fine))

        color = mode_colors[k]
        label_num = f'$n = {n_mode}$' if k < 3 else None
        label_exact = 'Exact' if k == 0 else None

        ax1.plot(x_fine, u_exact_fine, '--', color=color, linewidth=1.0,
                 alpha=0.6, label=label_exact)
        ax1.plot(x_phys, u_num, 'o', color=color, markersize=4,
                 markeredgecolor='white', markeredgewidth=0.3,
                 label=label_num)

    ax1.set_xlabel(r'$x$')
    ax1.set_ylabel(r'$u_n(x)$')
    ax1.set_title(f'First {n_modes_plot} Eigenmodes ($N = {N}$)', fontsize=11)
    ax1.legend(loc='lower left', fontsize=8, ncol=2, framealpha=0.9)
    ax1.set_xlim(-0.02, 1.02)
    ax1.set_ylim(-1.3, 1.3)
    ax1.spines['top'].set_visible(False)
    ax1.spines['right'].set_visible(False)
    ax1.axhline(0, color='#888888', linewidth=0.5)
    ax1.grid(True, alpha=0.2)

    # ---- Right panel: eigenvalue errors ----
    ax2 = axes[1]

    mode_numbers = np.arange(1, n_modes_err + 1)
    errors = []

    for k in range(n_modes_err):
        n_mode = k + 1
        lam_exact = exact_eigenvalue(n_mode)
        lam_num = eigenvalues[k]
        err = max(np.abs(lam_num - lam_exact), 1e-16)
        errors.append(err)

    errors = np.array(errors)

    ax2.semilogy(mode_numbers, errors, 'o', color=CORAL, markersize=5,
                 markeredgecolor='white', markeredgewidth=0.5)

    # Resolution limit line at ~ 2N/3
    n_res = int(2 * N / 3)
    ax2.axvline(n_res, color=PURPLE, linestyle=':', linewidth=1.0, alpha=0.7)
    ax2.text(n_res + 0.5, 1e-2, r'$\approx 2N/3$', fontsize=9,
             color=PURPLE, va='center')

    # Machine precision line
    ax2.axhline(2.2e-16, color='gray', linestyle=':', linewidth=1, alpha=0.7)
    ax2.text(n_modes_err - 0.5, 5e-16, 'Machine\nprecision', fontsize=8,
             color='gray', ha='right', va='bottom')

    ax2.set_xlabel(r'Mode number $n$')
    ax2.set_ylabel(r'$|\lambda_n^{\mathrm{num}} - \lambda_n^{\mathrm{exact}}|$')
    ax2.set_title('Eigenvalue Errors', fontsize=11)
    ax2.set_xlim(0, n_modes_err + 1)
    ax2.set_ylim(1e-16, 1e6)
    ax2.grid(True, which='both', alpha=0.3)
    ax2.spines['top'].set_visible(False)
    ax2.spines['right'].set_visible(False)

    # Main title
    fig.suptitle(r"Vibrating String: $-u'' = \lambda u$, $u(0) = 0$, $u'(1) = 0$"
                 f" ($N = {N}$)",
                 fontsize=13, y=0.98)

    plt.tight_layout(rect=[0, 0, 1, 0.93])

    # Save figure
    output_file = OUTPUT_DIR / 'vibrating_string.pdf'
    fig.savefig(output_file, bbox_inches='tight', pad_inches=0.05)
    fig.savefig(output_file.with_suffix('.png'), bbox_inches='tight',
                pad_inches=0.05, dpi=300)

    print(f'Figure saved to: {output_file.resolve()}')
    plt.close(fig)

    # =========================================================================
    # Print eigenvalue table
    # =========================================================================
    print(f"\nEigenvalue Accuracy for N = {N}:")
    print("-" * 60)
    print(f"{'n':>4}  {'lambda_exact':>16}  {'lambda_num':>16}  {'Error':>12}")
    print("-" * 60)
    for k in range(min(20, len(eigenvalues))):
        n_mode = k + 1
        lam_exact = exact_eigenvalue(n_mode)
        lam_num = eigenvalues[k]
        err = np.abs(lam_num - lam_exact)
        print(f"{n_mode:>4d}  {lam_exact:>16.8f}  {lam_num:>16.8f}  {err:>12.4e}")
    print("-" * 60)
    print(f"\nResolution limit: ~2N/3 = {2*N//3} well-resolved modes")
    print(f"Total positive eigenvalues found: {len(eigenvalues)}")


if __name__ == '__main__':
    main()

File: codes/julia/ch13/bc_vibrating_string.jl
# bc_vibrating_string.jl
#
# Chapter 13: Boundary Conditions in Spectral Methods
#
# Vibrating string eigenvalue problem with mixed boundary conditions.
#
#     -u_xx = lambda * u   on [0, 1]
#     u(0) = 0             (Dirichlet at left end)
#     u'(1) = 0            (Neumann / free end at right)
#
# The domain [0, 1] is mapped to the Chebyshev domain [-1, 1] via x = (xi + 1)/2,
# so D_x = 2 * D_xi.
#
# Chebyshev ordering: xi[1] = 1 -> physical x = 1 (free end, Neumann)
#                     xi[N+1] = -1 -> physical x = 0 (fixed end, Dirichlet)
#
# The operator A = -D_phys^2 with boundary rows replaced:
#     Row 1:   D_phys[1,:] (Neumann condition u'(1) = 0)
#     Row N+1: identity row (Dirichlet condition u(0) = 0)
#
# Exact eigenvalues: lambda_n = ((2n-1)*pi/2)^2,  n = 1, 2, ...
# Exact eigenfunctions: u_n(x) = sin((2n-1)*pi*x/2)
#
# Generates Figure 13.10: First 6 eigenmodes and eigenvalue errors.
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LinearAlgebra
using Printf

# Import Chebyshev matrix function from Chapter 7
include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

# Publication-quality CairoMakie configuration
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# Book color scheme
const NAVY   = colorant"#142D6E"
const SKY    = colorant"#7896D2"
const CORAL  = colorant"#E74C3C"
const TEAL   = colorant"#16A085"
const PURPLE = colorant"#8E44AD"
const ORANGE = colorant"#E67E22"

# Output paths
const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch13", "julia")

# -----------------------------------------------------------------------------
# Eigenvalue problem solver
# -----------------------------------------------------------------------------
"""
    solve_vibrating_string(N)

Solve the vibrating string eigenvalue problem:
    -u_xx = lambda * u on [0,1], u(0) = 0, u'(1) = 0

Returns sorted eigenvalues and eigenvectors on the physical grid x in [0, 1].
"""
function solve_vibrating_string(N)
    D, xi = cheb_matrix(N)
    n = N + 1

    # Map to physical domain [0, 1]: x = (xi + 1)/2
    # D_x = 2 * D_xi
    D_phys = 2.0 * D
    D2_phys = D_phys * D_phys
    x_phys = (xi .+ 1.0) ./ 2.0  # x_phys[1]=1 (free), x_phys[N+1]=0 (fixed)

    # Build operator A = -D2_phys
    A = -D2_phys

    # Row 1 (xi=1 -> x=1): Neumann u'(1) = 0
    A[1, :] = D_phys[1, :]

    # Row N+1 (xi=-1 -> x=0): Dirichlet u(0) = 0
    A[n, :] .= 0.0
    A[n, n] = 1.0

    # Solve eigenvalue problem
    F = eigen(A)
    lambdas = real.(F.values)
    vecs = real.(F.vectors)

    # Sort by eigenvalue magnitude
    idx = sortperm(lambdas)
    lambdas_sorted = lambdas[idx]
    vecs_sorted = vecs[:, idx]

    return lambdas_sorted, vecs_sorted, x_phys
end

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    mkpath(OUTPUT_DIR)

    N = 36
    lambdas, vecs, x_phys = solve_vibrating_string(N)

    # Exact eigenvalues: lambda_n = ((2n-1)*pi/2)^2
    n_exact = 20
    exact_lambdas = [((2*n - 1) * pi / 2)^2 for n in 1:n_exact]

    # Fine grid for exact eigenfunctions
    x_fine = collect(range(0, 1, length=200))

    # =========================================================================
    # Figure: Two panels
    # =========================================================================
    fig = Figure(size = (900, 420))

    # ---- Panel 1: First 6 eigenmodes ----
    ax1 = Axis(fig[1, 1],
               xlabel = L"x",
               ylabel = L"u_n(x)",
               title = "First 6 eigenmodes")

    colors_modes = [NAVY, TEAL, CORAL, PURPLE, ORANGE, SKY]
    n_modes = 6

    # Skip spurious eigenvalues near zero (from boundary rows)
    # Find the first positive eigenvalue that matches exact
    # The first few sorted eigenvalues include 0s from boundary conditions
    # Filter: keep eigenvalues > 1.0 (first exact is ~2.47)
    valid_mask = lambdas .> 1.0
    valid_idx = findall(valid_mask)

    for mode in 1:n_modes
        if mode > length(valid_idx)
            break
        end
        idx = valid_idx[mode]
        lam_computed = lambdas[idx]
        lam_exact = exact_lambdas[mode]

        # Computed eigenfunction
        psi = vecs[:, idx]
        # Normalize: max |psi| = 1
        psi ./= maximum(abs.(psi))
        # Ensure consistent sign (positive at first interior max)
        if psi[N ÷ 2] < 0
            psi .*= -1
        end

        # Exact eigenfunction
        u_exact = sin.((2 * mode - 1) .* pi .* x_fine ./ 2.0)
        u_exact ./= maximum(abs.(u_exact))

        # Sort physical grid for plotting
        idx_sort = sortperm(x_phys)

        # Plot exact (thin dashed) and computed (solid with markers)
        lines!(ax1, x_fine, u_exact, color = (colors_modes[mode], 0.4),
               linewidth = 1.0, linestyle = :dash)
        scatterlines!(ax1, x_phys[idx_sort], psi[idx_sort],
                       color = colors_modes[mode], linewidth = 1.5,
                       markersize = 3, markercolor = colors_modes[mode],
                       label = @sprintf("n=%d (λ=%.2f)", mode, lam_computed))
    end

    xlims!(ax1, -0.02, 1.02)
    axislegend(ax1, position = :lb, framevisible = true, labelsize = 8, nbanks = 2)
    hidespines!(ax1, :t, :r)

    # ---- Panel 2: Eigenvalue errors ----
    ax2 = Axis(fig[1, 2],
               xlabel = "Mode number n",
               ylabel = L"|\lambda_n - \lambda_n^{\mathrm{exact}}|",
               title = @sprintf("Eigenvalue errors (N = %d)", N),
               yscale = log10)

    n_compare = min(length(valid_idx), n_exact)
    errors = Float64[]
    mode_numbers = Int[]

    for mode in 1:n_compare
        idx = valid_idx[mode]
        err = abs(lambdas[idx] - exact_lambdas[mode])
        err = max(err, 1e-16)
        push!(errors, err)
        push!(mode_numbers, mode)
    end

    scatterlines!(ax2, mode_numbers, errors, color = NAVY, linewidth = 1.5,
                  markersize = 6, markercolor = NAVY,
                  strokecolor = :white, strokewidth = 0.5)

    hidespines!(ax2, :t, :r)
    hlines!(ax2, [2.2e-16], color = :gray, linestyle = :dot, linewidth = 1)
    text!(ax2, n_compare, 4e-16, text = "Machine precision", fontsize = 9,
          color = :gray, align = (:right, :bottom))

    # Main title
    Label(fig[0, :],
          L"Vibrating String: $-u_{xx} = \lambda u$, $u(0)=0$, $u'(1)=0$ ($N = %$(N)$)",
          fontsize = 13)

    # Save figure
    save(joinpath(OUTPUT_DIR, "vibrating_string.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "vibrating_string.png"), fig, px_per_unit = 4)
    println("Figure saved to: $(joinpath(OUTPUT_DIR, "vibrating_string.pdf"))")

    # =========================================================================
    # Print summary
    # =========================================================================
    println("\nVibrating String Eigenvalue Problem")
    println("="^60)
    @printf("  N = %d, domain [0, 1]\n", N)
    println("  BCs: u(0) = 0 (Dirichlet), u'(1) = 0 (Neumann)")
    println("  Exact eigenvalues: lambda_n = ((2n-1)*pi/2)^2")
    println()
    println("-"^60)
    @printf("%6s %16s %16s %12s\n", "Mode", "Computed", "Exact", "Error")
    println("-"^60)
    for mode in 1:min(10, n_compare)
        idx = valid_idx[mode]
        @printf("%6d %16.10f %16.10f %12.2e\n",
                mode, lambdas[idx], exact_lambdas[mode],
                abs(lambdas[idx] - exact_lambdas[mode]))
    end
    println("-"^60)
end

main()
